// @react/react-spectrum/Asset/js/Asset.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _AssetFile = require('./AssetFile');

var _AssetFile2 = _interopRequireDefault(_AssetFile);

var _AssetFolder = require('./AssetFolder');

var _AssetFolder2 = _interopRequireDefault(_AssetFolder);

var _AssetImage = require('./AssetImage');

var _AssetImage2 = _interopRequireDefault(_AssetImage);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require('../../spectrum-css-overrides/asset/index.css');

  require('../../spectrum-css-overrides/asset/index-diff.css');
} else if (process.env.SCALE_LARGE) {
  require('../../spectrum-css-overrides/asset/index-lg.css');
} else {
  require('../../spectrum-css-overrides/asset/index.css');
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require('../../spectrum-css-overrides/asset/multiStops/light.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require('../../spectrum-css-overrides/asset/multiStops/dark.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require('../../spectrum-css-overrides/asset/multiStops/lightest.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require('../../spectrum-css-overrides/asset/multiStops/darkest.css');

var Asset = function (_React$Component) {
  _inherits(Asset, _React$Component);

  function Asset() {
    _classCallCheck(this, Asset);

    return _possibleConstructorReturn(this, (Asset.__proto__ || Object.getPrototypeOf(Asset)).apply(this, arguments));
  }

  _createClass(Asset, [{
    key: 'render',
    value: function render(props) {
      var _props = this.props,
          type = _props.type,
          src = _props.src,
          placeholder = _props.placeholder,
          cache = _props.cache,
          headers = _props.headers,
          smartness = _props.smartness,
          alt = _props.alt,
          className = _props.className,
          onLoad = _props.onLoad,
          decorative = _props.decorative,
          otherProps = _objectWithoutProperties(_props, ['type', 'src', 'placeholder', 'cache', 'headers', 'smartness', 'alt', 'className', 'onLoad', 'decorative']);

      var content = void 0;

      if (type === 'image') {
        content = _react2.default.createElement(_AssetImage2.default, {
          src: src,
          placeholder: placeholder,
          smartness: smartness,
          onLoad: onLoad,
          cache: cache,
          headers: headers,
          alt: alt,
          decorative: decorative });
      } else if (type === 'folder') {
        content = _react2.default.createElement(_AssetFolder2.default, { alt: alt, decorative: decorative });
      } else {
        content = _react2.default.createElement(_AssetFile2.default, { alt: alt, decorative: decorative });
      }

      return _react2.default.createElement(
        'div',
        _extends({}, (0, _filterDOMProps2.default)(otherProps), {
          className: (0, _classnames2.default)('spectrum-Asset', className) }),
        content
      );
    }
  }]);

  return Asset;
}(_react2.default.Component);

Asset.propTypes = {
  /** Asset can be an image, folder or file type */
  type: _propTypes2.default.oneOf(['image', 'folder', 'file']),

  /** Source of the Asset if it is an image */
  src: _propTypes2.default.string,

  /** Whether to use the image cache for the asset image */
  cache: _propTypes2.default.bool,

  /** HTTP headers to add to the request for the asset image */
  headers: _propTypes2.default.object,

  /** Placeholder image to display while the fullsize one is loading, if cached. */
  placeholder: _propTypes2.default.string,

  /** A factor between 0 and 1 that controls how the preview is scaled. When 0 ("dumb") the assets fits all the space available. */
  smartness: _propTypes2.default.number,

  /** Alternate content for screen readers */
  alt: _propTypes2.default.string,

  /** Whether the image is being used for decoration and should not be announced by screen readers */
  decorative: _propTypes2.default.bool,

  /** Load callback triggered when images load */
  onLoad: _propTypes2.default.func
};
Asset.defaultProps = {
  type: 'file',
  smartness: 1
};
exports.default = Asset;
