// @react/react-spectrum/Grid/js/GridColumn.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _responsive = require('./responsive');

var _responsive2 = _interopRequireDefault(_responsive);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * A column in a grid
 * Responsive modifiers enable specifying different column sizes, offsets, and distribution at xs, sm, md & lg viewport widths.
 */
var GridColumn = function (_Component) {
  _inherits(GridColumn, _Component);

  function GridColumn() {
    _classCallCheck(this, GridColumn);

    return _possibleConstructorReturn(this, (GridColumn.__proto__ || Object.getPrototypeOf(GridColumn)).apply(this, arguments));
  }

  _createClass(GridColumn, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          className = _props.className,
          children = _props.children,
          size = _props.size,
          first = _props.first,
          last = _props.last,
          offsetSize = _props.offsetSize,
          otherProps = _objectWithoutProperties(_props, ['className', 'children', 'size', 'first', 'last', 'offsetSize']);

      var classes = (0, _classnames2.default)((0, _responsive2.default)('spectrum-grid-col-#size-#value', size), (0, _responsive2.default)('spectrum-grid-col-#size-offset-#value', offsetSize), (0, _responsive2.default)('spectrum-grid-first-#size', first), (0, _responsive2.default)('spectrum-grid-last-#size', last), className);

      return _react2.default.createElement(
        'div',
        _extends({ className: classes }, (0, _filterDOMProps2.default)(otherProps)),
        children
      );
    }
  }]);

  return GridColumn;
}(_react.Component);

GridColumn.propTypes = {
  /**
   * Custom classes to append to the grid column div
   */
  className: _propTypes2.default.string,

  /**
   * How many columns to span
   */
  size: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.oneOf(['auto']), _propTypes2.default.arrayOf(_propTypes2.default.number), _propTypes2.default.shape({
    xs: _propTypes2.default.number,
    sm: _propTypes2.default.number,
    md: _propTypes2.default.number,
    lg: _propTypes2.default.number,
    xl: _propTypes2.default.number
  })]),

  /**
   * Reorders element to first depending on viewport width
   */
  first: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['xs', 'sm', 'md', 'lg', 'xl']), _propTypes2.default.bool]),

  /**
   * Reorders element to last depending on viewport width
   */
  last: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['xs', 'sm', 'md', 'lg', 'xl']), _propTypes2.default.bool]),

  /**
   * How many columns over should the element start, useful for blank area
   */
  offsetSize: _propTypes2.default.oneOfType([_propTypes2.default.number, _propTypes2.default.arrayOf(_propTypes2.default.number), _propTypes2.default.shape({
    xs: _propTypes2.default.number,
    sm: _propTypes2.default.number,
    md: _propTypes2.default.number,
    lg: _propTypes2.default.number,
    xl: _propTypes2.default.number
  })])
};
GridColumn.defaultProps = {
  size: 'auto'
};
exports.default = GridColumn;
