// @react/react-spectrum/Grid/js/GridRow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _responsive = require('./responsive');

var _responsive2 = _interopRequireDefault(_responsive);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * A row in a grid
 */
var GridRow = function (_Component) {
  _inherits(GridRow, _Component);

  function GridRow() {
    _classCallCheck(this, GridRow);

    return _possibleConstructorReturn(this, (GridRow.__proto__ || Object.getPrototypeOf(GridRow)).apply(this, arguments));
  }

  _createClass(GridRow, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          align = _props.align,
          children = _props.children,
          valign = _props.valign,
          distribution = _props.distribution,
          reverse = _props.reverse,
          className = _props.className,
          otherProps = _objectWithoutProperties(_props, ['align', 'children', 'valign', 'distribution', 'reverse', 'className']);

      var classes = (0, _classnames2.default)('spectrum-grid-row', (0, _responsive2.default)('spectrum-grid-#value-#size', align), (0, _responsive2.default)('spectrum-grid-#value-#size', valign), (0, _responsive2.default)('spectrum-grid-#value-#size', distribution), { 'spectrum-grid-reverse': reverse }, className);

      return _react2.default.createElement(
        'div',
        _extends({ className: classes }, (0, _filterDOMProps2.default)(otherProps)),
        children
      );
    }
  }]);

  return GridRow;
}(_react.Component);

GridRow.propTypes = {
  /**
   * Fill me in and fix my prop type
   */
  align: _propTypes2.default.any,

  /**
   * Fill me in and fix my prop type
   */
  valign: _propTypes2.default.any,

  /**
   * Fill me in and fix my prop type
   */
  distribution: _propTypes2.default.any,

  /**
   * Fill me in and fix my prop type
   */
  reverse: _propTypes2.default.any,

  /**
   * Fill me in and fix my prop type
   */
  className: _propTypes2.default.string
};
GridRow.defaultProps = {
  reverse: false
};
exports.default = GridRow;
