// @react/react-spectrum/Card/js/CardBody.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Body of a card. Allows optional title, subtitle, and description props, and custom children displayed underneath.
 */
var CardBody = function (_React$Component) {
  _inherits(CardBody, _React$Component);

  function CardBody() {
    _classCallCheck(this, CardBody);

    return _possibleConstructorReturn(this, (CardBody.__proto__ || Object.getPrototypeOf(CardBody)).apply(this, arguments));
  }

  _createClass(CardBody, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          title = _props.title,
          subtitle = _props.subtitle,
          description = _props.description,
          children = _props.children;
      var _context = this.context,
          cardVariant = _context.cardVariant,
          actionMenu = _context.actionMenu;


      var subtitleWrapper = subtitle ? _react2.default.createElement(
        'div',
        { className: 'spectrum-Card-subtitle' },
        subtitle
      ) : null;
      var descriptionWrapper = description ? _react2.default.createElement(
        'div',
        { className: 'spectrum-Card-description' },
        description
      ) : null;

      return _react2.default.createElement(
        'div',
        { className: 'spectrum-Card-body' },
        title || actionMenu ? _react2.default.createElement(
          'div',
          { className: 'spectrum-Card-header' },
          title ? _react2.default.createElement(
            'div',
            { className: 'spectrum-Card-title' },
            title
          ) : null,
          cardVariant === 'gallery' && subtitleWrapper,
          cardVariant === 'gallery' && descriptionWrapper,
          actionMenu ? _react2.default.createElement(
            'div',
            { className: 'spectrum-Card-actionButton' },
            _react2.default.cloneElement(actionMenu, { alignRight: true })
          ) : null
        ) : null,
        cardVariant !== 'gallery' && (subtitle || description) ? _react2.default.createElement(
          'div',
          { className: 'spectrum-Card-content' },
          subtitleWrapper,
          descriptionWrapper
        ) : null,
        children
      );
    }
  }]);

  return CardBody;
}(_react2.default.Component);

CardBody.displayName = 'CardBody';
CardBody.propTypes = {
  /** Title for the card */
  title: _propTypes2.default.string,

  /** Subtitle for the card, e.g. file type, displayed in capitalized style */
  subtitle: _propTypes2.default.string,

  /** Description for the card */
  description: _propTypes2.default.string
};
CardBody.contextTypes = {
  cardVariant: _propTypes2.default.string,
  actionMenu: _propTypes2.default.element
};
exports.default = CardBody;
