// @react/react-spectrum/Menu/js/Menu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _List = require('../../List');

var _Popover = require('../../Popover');

var _Popover2 = _interopRequireDefault(_Popover);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Menu = function (_React$Component) {
  _inherits(Menu, _React$Component);

  function Menu() {
    _classCallCheck(this, Menu);

    return _possibleConstructorReturn(this, (Menu.__proto__ || Object.getPrototypeOf(Menu)).apply(this, arguments));
  }

  _createClass(Menu, [{
    key: 'getListRef',
    value: function getListRef() {
      return this.listRef;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          arrowStyle = _props.arrowStyle,
          children = _props.children,
          className = _props.className,
          open = _props.open,
          onClose = _props.onClose,
          onSelect = _props.onSelect,
          placement = _props.placement,
          _props$role = _props.role,
          role = _props$role === undefined ? 'menu' : _props$role,
          style = _props.style,
          otherProps = _objectWithoutProperties(_props, ['arrowStyle', 'children', 'className', 'open', 'onClose', 'onSelect', 'placement', 'role', 'style']);

      delete otherProps.target;

      return _react2.default.createElement(
        _Popover2.default,
        { arrowStyle: arrowStyle, isDialog: false, placement: placement, open: open, onClose: onClose, style: style },
        _react2.default.createElement(
          _List.List,
          _extends({ ref: function ref(l) {
              return _this2.listRef = l;
            }, role: role, className: className }, otherProps),
          _react2.default.Children.map(children, function (child) {
            return _react2.default.cloneElement(child, {
              onTab: function onTab(e) {
                return e.preventDefault();
              },
              onSelect: onSelect
            });
          })
        )
      );
    }
  }]);

  return Menu;
}(_react2.default.Component);

Menu.displayName = 'Menu';
Menu.propTypes = {
  placement: _propTypes2.default.oneOf(['bottom', 'bottom left', 'bottom right', 'top', 'top left', 'top right', 'left', 'left top', 'left bottom', 'right', 'right top', 'right bottom']),
  open: _propTypes2.default.bool,
  onClose: _propTypes2.default.func,
  onSelect: _propTypes2.default.func,
  className: _propTypes2.default.string
};
exports.default = Menu;
