// @react/react-spectrum/StepList/js/Step.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Step;

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _OverlayTrigger = require('../../OverlayTrigger');

var _OverlayTrigger2 = _interopRequireDefault(_OverlayTrigger);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Tooltip = require('../../Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var SMALL = 'S';

function Step(_ref) {
  var children = _ref.children,
      className = _ref.className,
      complete = _ref.complete,
      selected = _ref.selected,
      size = _ref.size,
      otherProps = _objectWithoutProperties(_ref, ['children', 'className', 'complete', 'selected', 'size']);

  return _react2.default.createElement(
    'div',
    _extends({
      className: (0, _classnames2.default)(className, 'spectrum-Steplist-item', { 'is-complete': complete }, { 'is-selected': selected }),
      role: 'tab',
      'aria-selected': selected
    }, (0, _filterDOMProps2.default)(otherProps)),
    _react2.default.createElement(
      'div',
      { className: 'spectrum-Steplist-label' },
      size !== SMALL && children
    ),
    _react2.default.createElement(
      'span',
      { className: 'spectrum-Steplist-markerContainer' },
      size === SMALL ? _react2.default.createElement(
        _OverlayTrigger2.default,
        { placement: 'top' },
        _react2.default.createElement('span', { className: 'spectrum-Steplist-marker' }),
        _react2.default.createElement(
          _Tooltip2.default,
          { open: true, className: 'spectrum-Steplist-tooltip' },
          children
        )
      ) : _react2.default.createElement('span', { className: 'spectrum-Steplist-marker' })
    ),
    _react2.default.createElement('span', { className: 'spectrum-Steplist-segment' })
  );
}

Step.displayName = 'Step';

Step.propTypes = {
  /** Class to add to the step */
  className: _propTypes2.default.string,

  /** Whether the current step is completed */
  complete: _propTypes2.default.bool,

  /** Whether the current step is selected */
  selected: _propTypes2.default.bool,

  /** Define the size of the step */
  size: _propTypes2.default.oneOf(['S', 'L'])
};
