// @react/react-spectrum/Alert/js/Alert.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Alert;

var _AlertMedium = require('../../Icon/core/AlertMedium');

var _AlertMedium2 = _interopRequireDefault(_AlertMedium);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _HelpMedium = require('../../Icon/core/HelpMedium');

var _HelpMedium2 = _interopRequireDefault(_HelpMedium);

var _InfoMedium = require('../../Icon/core/InfoMedium');

var _InfoMedium2 = _interopRequireDefault(_InfoMedium);

var _intl = require('../../utils/intl');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _SuccessMedium = require('../../Icon/core/SuccessMedium');

var _SuccessMedium2 = _interopRequireDefault(_SuccessMedium);

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var intlMessages = {
  'cs-CZ': _interopRequireDefault(require('./../intl/cs-CZ.json')).default,
  'da-DK': _interopRequireDefault(require('./../intl/da-DK.json')).default,
  'de-DE': _interopRequireDefault(require('./../intl/de-DE.json')).default,
  'en-US': _interopRequireDefault(require('./../intl/en-US.json')).default,
  'es-ES': _interopRequireDefault(require('./../intl/es-ES.json')).default,
  'fi-FI': _interopRequireDefault(require('./../intl/fi-FI.json')).default,
  'fr-FR': _interopRequireDefault(require('./../intl/fr-FR.json')).default,
  'it-IT': _interopRequireDefault(require('./../intl/it-IT.json')).default,
  'ja-JP': _interopRequireDefault(require('./../intl/ja-JP.json')).default,
  'ko-KR': _interopRequireDefault(require('./../intl/ko-KR.json')).default,
  'nb-NO': _interopRequireDefault(require('./../intl/nb-NO.json')).default,
  'nl-NL': _interopRequireDefault(require('./../intl/nl-NL.json')).default,
  'pl-PL': _interopRequireDefault(require('./../intl/pl-PL.json')).default,
  'pt-BR': _interopRequireDefault(require('./../intl/pt-BR.json')).default,
  'ru-RU': _interopRequireDefault(require('./../intl/ru-RU.json')).default,
  'sv-SE': _interopRequireDefault(require('./../intl/sv-SE.json')).default,
  'tr-TR': _interopRequireDefault(require('./../intl/tr-TR.json')).default,
  'zh-CN': _interopRequireDefault(require('./../intl/zh-CN.json')).default,
  'zh-TW': _interopRequireDefault(require('./../intl/zh-TW.json')).default
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require('../../spectrum-css/alert/index.css');

  require('../../spectrum-css/alert/index-diff.css');
} else if (process.env.SCALE_LARGE) {
  require('../../spectrum-css/alert/index-lg.css');
} else {
  require('../../spectrum-css/alert/index.css');
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require('../../spectrum-css/alert/multiStops/light.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require('../../spectrum-css/alert/multiStops/dark.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require('../../spectrum-css/alert/multiStops/lightest.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require('../../spectrum-css/alert/multiStops/darkest.css');


var ICONS = {
  error: _AlertMedium2.default,
  warning: _AlertMedium2.default,
  info: _InfoMedium2.default,
  help: _HelpMedium2.default,
  success: _SuccessMedium2.default
};

var formatMessage = (0, _intl.messageFormatter)(intlMessages);

function Alert(_ref) {
  var header = _ref.header,
      children = _ref.children,
      _ref$variant = _ref.variant,
      variant = _ref$variant === undefined ? 'info' : _ref$variant,
      className = _ref.className,
      _ref$alt = _ref.alt,
      alt = _ref$alt === undefined ? formatMessage(variant || 'info') : _ref$alt,
      otherProps = _objectWithoutProperties(_ref, ['header', 'children', 'variant', 'className', 'alt']);

  var AlertIcon = ICONS[variant];

  return _react2.default.createElement(
    'div',
    _extends({
      className: (0, _classnames2.default)('spectrum-Alert', 'spectrum-Alert--' + variant, className),
      role: 'alert'
    }, (0, _filterDOMProps2.default)(otherProps)),
    _react2.default.createElement(AlertIcon, { size: null, className: 'spectrum-Alert-icon', alt: alt }),
    _react2.default.createElement(
      'div',
      { className: 'spectrum-Alert-header' },
      header
    ),
    _react2.default.createElement(
      'div',
      { className: 'spectrum-Alert-content' },
      children
    )
  );
}

Alert.propTypes = {
  /**
   * A string for the Header of the Alert
   */
  header: _propTypes2.default.string,

  /**
   * Any arbitrary node to render into content area
   */
  children: _propTypes2.default.node,

  /**
   * Affects the color and icon used by the Alert
   */
  variant: _propTypes2.default.oneOf(['error', 'warning', 'info', 'help', 'success'])
};

Alert.displayName = 'Alert';
