// @react/react-spectrum/Form/js/FormItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _LabelBase = require('../../FieldLabel/js/LabelBase');

var _LabelBase2 = _interopRequireDefault(_LabelBase);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FormItem = function (_Component) {
  _inherits(FormItem, _Component);

  function FormItem() {
    _classCallCheck(this, FormItem);

    return _possibleConstructorReturn(this, (FormItem.__proto__ || Object.getPrototypeOf(FormItem)).apply(this, arguments));
  }

  _createClass(FormItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          label = _props.label,
          labelAlign = _props.labelAlign,
          labelFor = _props.labelFor,
          className = _props.className,
          children = _props.children,
          otherProps = _objectWithoutProperties(_props, ['label', 'labelAlign', 'labelFor', 'className', 'children']);

      var labelClassNames = (0, _classnames2.default)('spectrum-Form-itemLabel', 'spectrum-FieldLabel--' + labelAlign);

      return _react2.default.createElement(
        _LabelBase2.default,
        _extends({
          label: label,
          className: (0, _classnames2.default)('spectrum-Form-item', className),
          labelClassName: labelClassNames,
          wrapperClassName: 'spectrum-Form-itemField',
          labelFor: labelFor,
          componentName: 'FormItem'
        }, otherProps),
        children
      );
    }
  }]);

  return FormItem;
}(_react.Component);

FormItem.propTypes = {
  /** Label for form field. Can be Text of HTML */
  label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.object]),
  /** Label Alignment. Defaults to left */
  labelAlign: _propTypes2.default.oneOf(['left', 'right']),
  /** Id of the labelable form element */
  labelFor: _propTypes2.default.string
};
FormItem.defaultProps = {
  labelAlign: 'left'
};
exports.default = FormItem;
