// @react/react-spectrum/IllustratedMessage/js/IllustratedMessage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _Heading = require('../../Heading');

var _Heading2 = _interopRequireDefault(_Heading);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require('../../spectrum-css/illustratedmessage/index.css');

  require('../../spectrum-css/illustratedmessage/index-diff.css');
} else if (process.env.SCALE_LARGE) {
  require('../../spectrum-css/illustratedmessage/index-lg.css');
} else {
  require('../../spectrum-css/illustratedmessage/index.css');
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require('../../spectrum-css/illustratedmessage/multiStops/light.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require('../../spectrum-css/illustratedmessage/multiStops/dark.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require('../../spectrum-css/illustratedmessage/multiStops/lightest.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require('../../spectrum-css/illustratedmessage/multiStops/darkest.css');

/**
 * An IllustratedMessage displays an illustration icon and a message, usually in an empty state or on an error page.
 */
var IllustratedMessage = function (_React$Component) {
  _inherits(IllustratedMessage, _React$Component);

  function IllustratedMessage() {
    _classCallCheck(this, IllustratedMessage);

    return _possibleConstructorReturn(this, (IllustratedMessage.__proto__ || Object.getPrototypeOf(IllustratedMessage)).apply(this, arguments));
  }

  _createClass(IllustratedMessage, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          _props$illustration = _props.illustration,
          illustration = _props$illustration === undefined ? null : _props$illustration,
          otherProps = _objectWithoutProperties(_props, ['illustration']);

      if (illustration && (!illustration.props.className || !illustration.props.className.includes('spectrum-IllustratedMessage-illustration'))) {
        illustration = _react2.default.cloneElement(illustration, {
          className: (0, _classnames2.default)(illustration.props.className, 'spectrum-IllustratedMessage-illustration')
        });
      }

      return _react2.default.createElement(
        'div',
        _extends({}, (0, _filterDOMProps2.default)(otherProps), {
          className: (0, _classnames2.default)('spectrum-IllustratedMessage', this.props.className) }),
        illustration,
        this.props.heading && _react2.default.createElement(
          _Heading2.default,
          { size: 2, className: 'spectrum-IllustratedMessage-heading' },
          this.props.heading
        ),
        this.props.description && _react2.default.createElement(
          'p',
          { className: 'spectrum-Body--secondary spectrum-IllustratedMessage-description' },
          this.props.description
        )
      );
    }
  }]);

  return IllustratedMessage;
}(_react2.default.Component);

IllustratedMessage.propTypes = {
  /** The heading to be displayed */
  heading: _propTypes2.default.string,

  /** The description to be displayed */
  description: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.element]),

  /** The illustration to be rendered above heading. Should be an SVG element. */
  illustration: _propTypes2.default.element
};
IllustratedMessage.defaultProps = {
  heading: '',
  description: ''
};
exports.default = IllustratedMessage;
