// @react/react-spectrum/Shell/js/ShellSolutionSwitcher.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = ShellSolutionSwitcher;

var _Apps = require('../../Icon/Apps');

var _Apps2 = _interopRequireDefault(_Apps);

var _Button = require('../../Button');

var _Button2 = _interopRequireDefault(_Button);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ShellMenu = require('./ShellMenu');

var _ShellMenu2 = _interopRequireDefault(_ShellMenu);

require('../style/ShellSolutionSwitcher.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function ShellSolutionSwitcher(_ref) {
  var className = _ref.className,
      children = _ref.children,
      otherProps = _objectWithoutProperties(_ref, ['className', 'children']);

  return _react2.default.createElement(
    _ShellMenu2.default,
    _extends({
      placement: 'top',
      animateFrom: 'top',
      full: true,
      dark: true,
      top: true,
      target: _react2.default.createElement(_Button2.default, {
        className: 'coral3-Shell-menu-button',
        variant: 'minimal',
        icon: _react2.default.createElement(_Apps2.default, null) })
    }, otherProps),
    _react2.default.createElement(
      'div',
      {
        className: (0, _classnames2.default)('coral3-Shell-solutionSwitcher', className) },
      _react2.default.createElement(
        'div',
        { className: 'coral3-Shell-solutionSwitcher-content' },
        children
      )
    )
  );
}

ShellSolutionSwitcher.displayName = 'ShellSolutionSwitcher';
