// @react/react-spectrum/Shell/js/ShellUserProfile.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = ShellUserProfile;

var _Button = require('../../Button');

var _Button2 = _interopRequireDefault(_Button);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ShellMenu = require('./ShellMenu');

var _ShellMenu2 = _interopRequireDefault(_ShellMenu);

require('../style/ShellUserProfile.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function ShellUserProfile(_ref) {
  var name = _ref.name,
      heading = _ref.heading,
      subheading = _ref.subheading,
      _ref$avatarUrl = _ref.avatarUrl,
      avatarUrl = _ref$avatarUrl === undefined ? '#' : _ref$avatarUrl,
      _ref$profileUrl = _ref.profileUrl,
      profileUrl = _ref$profileUrl === undefined ? '#' : _ref$profileUrl,
      _ref$onSignOut = _ref.onSignOut,
      onSignOut = _ref$onSignOut === undefined ? function () {} : _ref$onSignOut,
      children = _ref.children,
      className = _ref.className,
      otherProps = _objectWithoutProperties(_ref, ['name', 'heading', 'subheading', 'avatarUrl', 'profileUrl', 'onSignOut', 'children', 'className']);

  return _react2.default.createElement(
    _ShellMenu2.default,
    _extends({
      placement: 'right',
      animateFrom: 'top',
      target: _react2.default.createElement(
        _Button2.default,
        {
          className: 'coral3-Shell-menu-button',
          variant: 'minimal' },
        _react2.default.createElement('img', { src: avatarUrl, alt: '', 'aria-hidden': 'true' })
      )
    }, otherProps),
    _react2.default.createElement(
      'div',
      {
        className: (0, _classnames2.default)('coral3-Shell-user', className) },
      _react2.default.createElement(
        'div',
        { className: 'coral3-Shell-user-container' },
        _react2.default.createElement(
          'div',
          { className: 'coral3-Shell-user-image' },
          _react2.default.createElement('img', { src: avatarUrl, alt: '', 'aria-hidden': 'true', className: 'coral3-Shell-user-avatar' })
        ),
        _react2.default.createElement(
          'div',
          { className: 'coral3-Shell-user-name' },
          name
        ),
        _react2.default.createElement(
          'div',
          { className: 'coral3-Shell-user-heading' },
          heading && heading !== 'null' ? heading : ''
        ),
        _react2.default.createElement(
          'div',
          { className: 'coral3-Shell-user-subheading' },
          subheading && subheading !== 'null' ? subheading : ''
        )
      ),
      _react2.default.createElement(
        'div',
        { className: 'coral3-Shell-user-content' },
        children
      ),
      _react2.default.createElement(
        'div',
        { className: 'coral3-Shell-user-footer' },
        _react2.default.createElement(
          _Button2.default,
          { element: 'a', variant: 'minimal', href: profileUrl },
          'Edit Profile'
        ),
        _react2.default.createElement(
          _Button2.default,
          { element: 'a', variant: 'minimal', onClick: onSignOut },
          'Sign Out'
        )
      )
    )
  );
}

ShellUserProfile.displayName = 'ShellUserProfile';
