// @react/react-spectrum/Shell/js/ShellHelp.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Button = require('../../Button');

var _Button2 = _interopRequireDefault(_Button);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _HelpOutline = require('../../Icon/HelpOutline');

var _HelpOutline2 = _interopRequireDefault(_HelpOutline);

var _Link = require('../../Link');

var _Link2 = _interopRequireDefault(_Link);

var _List = require('../../List');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Search = require('../../Search');

var _Search2 = _interopRequireDefault(_Search);

var _ShellMenu = require('./ShellMenu');

var _ShellMenu2 = _interopRequireDefault(_ShellMenu);

var _Wait = require('../../Wait');

var _Wait2 = _interopRequireDefault(_Wait);

require('../style/ShellHelp.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ShellHelp = function (_Component) {
  _inherits(ShellHelp, _Component);

  function ShellHelp() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ShellHelp);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ShellHelp.__proto__ || Object.getPrototypeOf(ShellHelp)).call.apply(_ref, [this].concat(args))), _this), _initialiseProps.call(_this), _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ShellHelp, [{
    key: 'renderResults',
    value: function renderResults() {
      var searchResults = this.props.searchResults;

      if (searchResults) {
        return this.renderSearchResults();
      }
      return this.renderDefaultResults();
    }
  }, {
    key: 'renderSearchResults',
    value: function renderSearchResults() {
      var _this2 = this;

      var _props = this.props,
          searchResults = _props.searchResults,
          numTotalResults = _props.numTotalResults,
          moreSearchResultsUrl = _props.moreSearchResultsUrl;
      var searchTerm = this.state.searchTerm;

      // https://git.corp.adobe.com/React/react-coral/issues/134
      // unicode for: '&nbsp; &bull; &nbsp';

      var separator = '\xA0 \u2022 \xA0';
      var allResults = {
        href: moreSearchResultsUrl + '?q=' + searchTerm,
        title: 'See all ' + numTotalResults + ' results'
      };

      return _react2.default.createElement(
        _List.List,
        { className: 'coral3-Shell-help-results' },
        numTotalResults !== 0 && searchResults && searchResults.length ? searchResults.map(function (result, index) {
          return _react2.default.createElement(
            _List.ListItem,
            {
              key: index,
              onClick: _this2.handleResultClick.bind(_this2, result) },
            _react2.default.createElement(
              _Link2.default,
              {
                href: result.href,
                target: '_blank',
                rel: 'noopener noreferrer' },
              _react2.default.createElement(
                'span',
                null,
                result.title,
                _react2.default.createElement(
                  'div',
                  { className: 'coral3-Shell-help-result-description' },
                  Array.isArray(result.tags) ? result.tags.join(separator) : ''
                )
              )
            )
          );
        }).concat(_react2.default.createElement(
          _List.ListItem,
          {
            key: 'all-results',
            onClick: this.handleResultClick.bind(this, allResults) },
          _react2.default.createElement(
            _Link2.default,
            {
              className: 'spectrum-Link coral3-Shell-help-allResults',
              href: allResults.href,
              target: '_blank',
              rel: 'noopener noreferrer' },
            allResults.title
          )
        )) : _react2.default.createElement(
          'div',
          { className: 'coral3-Shell-help-resultMessage' },
          _react2.default.createElement(
            'div',
            { className: 'coral3-Shell-help-resultMessage-container' },
            _react2.default.createElement(
              'div',
              { className: 'coral-Heading--1 coral3-Shell-help-resultMessage-heading' },
              'No results found'
            )
          )
        )
      );
    }
  }, {
    key: 'renderDefaultResults',
    value: function renderDefaultResults() {
      var _this3 = this;

      var defaultResults = this.props.defaultResults;


      return _react2.default.createElement(
        _List.List,
        { className: 'coral3-Shell-help-items' },
        defaultResults && defaultResults.length ? defaultResults.map(function (result, index) {
          return _react2.default.createElement(
            _List.ListItem,
            {
              icon: result.icon,
              key: index,
              onClick: _this3.handleResultClick.bind(_this3, result) },
            _react2.default.createElement(
              _Link2.default,
              {
                className: 'coral3-Shell-help-item',
                href: result.href,
                target: '_blank',
                rel: 'noopener noreferrer' },
              result.label
            )
          );
        }) : null
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _this4 = this;

      var _props2 = this.props,
          className = _props2.className,
          loading = _props2.loading,
          otherProps = _objectWithoutProperties(_props2, ['className', 'loading']);

      return _react2.default.createElement(
        _ShellMenu2.default,
        _extends({
          placement: 'right',
          animateFrom: 'top',
          target: _react2.default.createElement(_Button2.default, {
            className: 'coral3-Shell-menu-button',
            variant: 'minimal',
            icon: _react2.default.createElement(_HelpOutline2.default, null) }),
          onVisible: this.handleVisible,
          onHidden: this.props.onHide
        }, otherProps),
        _react2.default.createElement(
          'div',
          {
            ref: function ref(el) {
              _this4.contentRef = el;
            },
            className: (0, _classnames2.default)('coral3-Shell-help', className) },
          _react2.default.createElement(_Search2.default, {
            className: 'coral3-Shell-help-search',
            placeholder: 'Search Help',
            onSubmit: this.handleSearch,
            onChange: this.props.onChange,
            quiet: true }),
          loading ? _react2.default.createElement(
            'div',
            { className: 'coral3-Shell-help-loading' },
            _react2.default.createElement(_Wait2.default, { className: 'coral3-Shell-help-loading-wait' }),
            _react2.default.createElement(
              'span',
              { className: 'coral-Heading--2 coral3-Shell-help-loading-info' },
              'Searching Help...'
            )
          ) : this.renderResults()
        )
      );
    }
  }]);

  return ShellHelp;
}(_react.Component);

ShellHelp.defaultProps = {
  defaultResults: [],
  onSearch: function onSearch() {},
  onChange: function onChange() {},
  onHide: function onHide() {},
  onResultClick: function onResultClick() {}
};

var _initialiseProps = function _initialiseProps() {
  var _this5 = this;

  this.state = {
    searchTerm: ''
  };

  this.handleSearch = function (val) {
    var onSearch = _this5.props.onSearch;

    onSearch(val);
    _this5.setState({
      searchTerm: val
    });
  };

  this.handleVisible = function () {
    if (_this5.contentRef) {
      _this5.contentRef.querySelector('.spectrum-Search-input').focus();
    }
  };

  this.handleResultClick = function () {
    if (_this5.props.onResultClick) {
      var _props3;

      (_props3 = _this5.props).onResultClick.apply(_props3, arguments);
    }
  };
};

exports.default = ShellHelp;


ShellHelp.displayName = 'ShellHelp';
