// @react/react-spectrum/Rating/js/Rating.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _autobindDecorator = require('autobind-decorator');

var _autobindDecorator2 = _interopRequireDefault(_autobindDecorator);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _createId = require('../../utils/createId');

var _createId2 = _interopRequireDefault(_createId);

var _filterDOMProps = require('../../utils/filterDOMProps');

var _filterDOMProps2 = _interopRequireDefault(_filterDOMProps);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Star = require('../../Icon/core/Star');

var _Star2 = _interopRequireDefault(_Star);

var _StarOutline = require('../../Icon/core/StarOutline');

var _StarOutline2 = _interopRequireDefault(_StarOutline);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require('../../spectrum-css/rating/index.css');

  require('../../spectrum-css/rating/index-diff.css');
} else if (process.env.SCALE_LARGE) {
  require('../../spectrum-css/rating/index-lg.css');
} else {
  require('../../spectrum-css/rating/index.css');
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require('../../spectrum-css/rating/multiStops/light.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require('../../spectrum-css/rating/multiStops/dark.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require('../../spectrum-css/rating/multiStops/lightest.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require('../../spectrum-css/rating/multiStops/darkest.css');

var Rating = (0, _autobindDecorator2.default)(_class = function (_React$Component) {
  _inherits(Rating, _React$Component);

  function Rating(props) {
    _classCallCheck(this, Rating);

    var _this = _possibleConstructorReturn(this, (Rating.__proto__ || Object.getPrototypeOf(Rating)).call(this, props));

    _this.state = {
      currentRating: _this.props.value || 0,
      currentFocus: null
    };

    _this.inputId = (0, _createId2.default)();
    return _this;
  }

  _createClass(Rating, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(props) {
      if (props.value != null) {
        this.setState({
          currentRating: props.value || 0
        });
      }
    }
  }, {
    key: 'onClickRating',
    value: function onClickRating(currentRating, e) {
      e.stopPropagation();

      // Allow user to set rating to zero.
      if (currentRating === 1 && this.state.currentRating === 1) {
        currentRating = 0;
      }

      if (this.props.value == null) {
        this.setState({ currentRating: currentRating });
      }

      if (this.props.onChange) {
        this.props.onChange(currentRating);
      }

      if (this.input) {
        this.input.focus();
      }
    }
  }, {
    key: 'onInput',
    value: function onInput(e) {
      var currentRating = +e.target.value;

      if (currentRating !== this.state.currentRating) {
        this.onClickRating(currentRating, e);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          max = _props.max,
          disabled = _props.disabled,
          className = _props.className,
          _props$id = _props.id,
          id = _props$id === undefined ? this.inputId : _props$id,
          otherProps = _objectWithoutProperties(_props, ['max', 'disabled', 'className', 'id']);

      var currentRating = this.state.currentRating;

      var ratings = [];

      for (var i = 1; i <= max; ++i) {
        var active = i <= Math.round(currentRating);

        ratings.push(_react2.default.createElement(
          'span',
          {
            key: i,
            className: (0, _classnames2.default)('spectrum-Rating-icon', { 'is-selected': active, 'is-disabled': disabled }),
            onClick: !disabled ? this.onClickRating.bind(this, i) : null,
            onKeyDown: !disabled ? function () {} : null },
          _react2.default.createElement(_Star2.default, { size: null, className: 'spectrum-Rating-starActive' }),
          _react2.default.createElement(_StarOutline2.default, { size: null, className: 'spectrum-Rating-starInactive' })
        ));
      }

      return _react2.default.createElement(
        'div',
        {
          className: (0, _classnames2.default)('spectrum-Rating', { 'is-disabled': disabled }, className) },
        _react2.default.createElement('input', _extends({
          ref: function ref(i) {
            return _this2.input = i;
          },
          id: id,
          className: 'spectrum-Rating-input',
          type: 'range',
          min: 0,
          max: max,
          value: currentRating,
          style: { width: 24 * max + 'px' },
          disabled: disabled || null,
          onInput: !disabled ? this.onInput.bind(this) : null
        }, (0, _filterDOMProps2.default)(otherProps))),
        ratings
      );
    }
  }]);

  return Rating;
}(_react2.default.Component)) || _class;

Rating.defaultProps = {
  disabled: false,
  className: '',
  max: 5,
  readOnly: false
};
exports.default = Rating;
