// @react/react-spectrum/Search/js/Search.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class;

var _autobindDecorator = require('autobind-decorator');

var _autobindDecorator2 = _interopRequireDefault(_autobindDecorator);

var _Button = require('../../Button');

var _Button2 = _interopRequireDefault(_Button);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _icon = require('../../utils/icon');

var _CrossSmall = require('../../Icon/core/CrossSmall');

var _CrossSmall2 = _interopRequireDefault(_CrossSmall);

var _Magnifier = require('../../Icon/core/Magnifier');

var _Magnifier2 = _interopRequireDefault(_Magnifier);

var _intl = require('../../utils/intl');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Textfield = require('../../Textfield');

var _Textfield2 = _interopRequireDefault(_Textfield);

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var intlMessages = {
  'cs-CZ': _interopRequireDefault(require('./../intl/cs-CZ.json')).default,
  'da-DK': _interopRequireDefault(require('./../intl/da-DK.json')).default,
  'de-DE': _interopRequireDefault(require('./../intl/de-DE.json')).default,
  'en-US': _interopRequireDefault(require('./../intl/en-US.json')).default,
  'es-ES': _interopRequireDefault(require('./../intl/es-ES.json')).default,
  'fi-FI': _interopRequireDefault(require('./../intl/fi-FI.json')).default,
  'fr-FR': _interopRequireDefault(require('./../intl/fr-FR.json')).default,
  'it-IT': _interopRequireDefault(require('./../intl/it-IT.json')).default,
  'ja-JP': _interopRequireDefault(require('./../intl/ja-JP.json')).default,
  'ko-KR': _interopRequireDefault(require('./../intl/ko-KR.json')).default,
  'nb-NO': _interopRequireDefault(require('./../intl/nb-NO.json')).default,
  'nl-NL': _interopRequireDefault(require('./../intl/nl-NL.json')).default,
  'pl-PL': _interopRequireDefault(require('./../intl/pl-PL.json')).default,
  'pt-BR': _interopRequireDefault(require('./../intl/pt-BR.json')).default,
  'ru-RU': _interopRequireDefault(require('./../intl/ru-RU.json')).default,
  'sv-SE': _interopRequireDefault(require('./../intl/sv-SE.json')).default,
  'tr-TR': _interopRequireDefault(require('./../intl/tr-TR.json')).default,
  'zh-CN': _interopRequireDefault(require('./../intl/zh-CN.json')).default,
  'zh-TW': _interopRequireDefault(require('./../intl/zh-TW.json')).default
};

if (process.env.SCALE_MEDIUM && process.env.SCALE_LARGE) {
  require('../../spectrum-css/search/index.css');

  require('../../spectrum-css/search/index-diff.css');
} else if (process.env.SCALE_LARGE) {
  require('../../spectrum-css/search/index-lg.css');
} else {
  require('../../spectrum-css/search/index.css');
}

if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHT) require('../../spectrum-css/search/multiStops/light.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARK) require('../../spectrum-css/search/multiStops/dark.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_LIGHTEST) require('../../spectrum-css/search/multiStops/lightest.css');
if (!process.env.THEME_LIGHT && !process.env.THEME_DARK && !process.env.THEME_LIGHTEST && !process.env.THEME_DARKEST || process.env.THEME_DARKEST) require('../../spectrum-css/search/multiStops/darkest.css');

var formatMessage = (0, _intl.messageFormatter)(intlMessages);

var normalizeValue = function normalizeValue(value) {
  return value || '';
};

var Search = (0, _autobindDecorator2.default)(_class = function (_Component) {
  _inherits(Search, _Component);

  function Search(props) {
    _classCallCheck(this, Search);

    var _this = _possibleConstructorReturn(this, (Search.__proto__ || Object.getPrototypeOf(Search)).call(this, props));

    var value = props.value,
        defaultValue = props.defaultValue;


    _this.state = {
      value: normalizeValue(value || defaultValue)
    };
    return _this;
  }

  _createClass(Search, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(props) {
      if (normalizeValue(props.value) !== this.state.value) {
        this.setState({
          value: props.value
        });
      }
    }
  }, {
    key: 'handleTextKeyDown',
    value: function handleTextKeyDown(e) {
      var _props = this.props,
          onSubmit = _props.onSubmit,
          onKeyDown = _props.onKeyDown,
          disabled = _props.disabled;
      var value = this.state.value;

      var key = e.which;

      if (key === 13 || key === 27) {
        e.preventDefault();
      }

      if (disabled) {
        return;
      }

      if (key === 13) {
        onSubmit(value);
      }

      if (key === 27) {
        this.handleTextChange('', e, 'escapeKey');
      }

      if (onKeyDown) {
        onKeyDown(e);
      }
    }
  }, {
    key: 'handleTextChange',
    value: function handleTextChange(value, e) {
      var from = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'input';
      var _props2 = this.props,
          onChange = _props2.onChange,
          disabled = _props2.disabled;

      if (disabled || value === this.state.value) {
        return;
      }

      if (!('value' in this.props)) {
        this.setState({
          value: value
        });
      }

      onChange(value, e, { from: from });
    }
  }, {
    key: 'handleClearButtonClick',
    value: function handleClearButtonClick(e) {
      this.handleTextChange('', e, 'clearButton');
      // restore focus to the searchbox
      if (this.searchbox) {
        this.searchbox.focus();
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props3 = this.props,
          disabled = _props3.disabled,
          className = _props3.className,
          icon = _props3.icon,
          _props3$role = _props3.role,
          role = _props3$role === undefined ? 'search' : _props3$role,
          otherProps = _objectWithoutProperties(_props3, ['disabled', 'className', 'icon', 'role']);

      var value = this.state.value;


      return _react2.default.createElement(
        'div',
        {
          role: role,
          className: (0, _classnames2.default)('spectrum-Search', { 'is-disabled': disabled }, className) },
        _react2.default.createElement(_Textfield2.default, _extends({
          role: 'searchbox',
          ref: function ref(s) {
            return _this2.searchbox = s;
          },
          className: 'spectrum-Search-input',
          value: value,
          disabled: disabled
        }, otherProps, {
          onKeyDown: this.handleTextKeyDown,
          onChange: this.handleTextChange })),
        (0, _icon.cloneIcon)(icon, { className: 'spectrum-Search-icon', size: 'S' }),
        value !== '' && _react2.default.createElement(_Button2.default, {
          'aria-label': formatMessage('Clear search'),
          variant: 'clear',
          icon: _react2.default.createElement(_CrossSmall2.default, null),
          disabled: disabled,
          onClick: this.handleClearButtonClick })
      );
    }
  }]);

  return Search;
}(_react.Component)) || _class;

Search.defaultProps = {
  icon: _react2.default.createElement(_Magnifier2.default, null),
  onChange: function onChange() {},
  onSubmit: function onSubmit() {}
};
exports.default = Search;


Search.displayName = 'Search';
